extensions/amp-apester-media/0.1/amp-apester-media.js

```diff
@@ -0,0 +1,235 @@
+import {getLengthNumeral, isLayoutSizeDefined} from '../../../src/layout';
+import {isExperimentOn} from '../../../src/experiments';
+import {removeElement} from '../../../src/dom';
+import {vsyncFor} from '../../../src/vsync';
+import {xhrFor} from '../../../src/xhr';
+
+
+/** @const */
+const TAG = 'amp-apester-media';
+
+/**
+ * AMP Apester-media
+ */
+class AmpApesterMedia extends AMP.BaseElement {
+  /** @override */
+  preconnectCallback(onLayout) {
+    this.preconnect.url(this.displayBaseUrl_, onLayout);
+    this.preconnect.url(this.rendererBaseUrl_, onLayout);
+  }
+
+  /** @override */
+  isLayoutSupported(layout) {
+    return isLayoutSizeDefined(layout);
+  }
+
+  /** @override */
+  viewportCallback(inViewport) {
+    if (inViewport && !this.seen_) {
+      if (this.iframe_) {
+        dev().fine(TAG, 'media seen');
+        this.seen_ = true;
+        this.iframe_.contentWindow./*OK*/postMessage('interaction seen', '*');
+      }
+    }
+  }
+
+  /** @override */
+  buildCallback() {
+
+    // EXPERIMENT
+    user().assert(isExperimentOn(this.win, TAG), `Enable ${TAG} experiment`);
+    const width = this.element.getAttribute('width');
+    const height = this.element.getAttribute('height');
+
+    /**
+     *  @private @const {number}
+     *  */
+    this.width_ = getLengthNumeral(width);
+
+    /**
+     * @private @const {number}
+     * */
+    this.height_ = getLengthNumeral(height);
+
+    /**
+     * @const @private {?string}
+     */
+    this.rendererBaseUrl_ = 'https://renderer.qmerce.com';
+
+    /**
+     * @const @private {?string}
+     */
+    this.displayBaseUrl_ = 'https://display.apester.com';
+
+    /**
+     * @private {boolean}
+     */
+    this.random = false;
+
+    /**
+     * @const @private {?string}
+     */
+    this.mediaAttribute_ = user().assert(
+        (this.element.getAttribute('data-apester-media-id') ||
+         (this.random =
+             this.element.getAttribute('data-apester-channel-token'))),
+        'Either the data-apester-media-id or the data-apester-channel-token ' +
+        'attributes must be specified for <amp-apester-media> %s',
+        this.element);
+
+    /**
+     * @private {?Element}
+     */
+    this.iframe_ = null;
+
+    /**
+     * @private {?Promise}
+     */
+    this.iframePromise_ = null;
+
+    /**
+     * @private {boolean}
+     */
+    this.seen_ = false;
+  }
+
+  /** @override */
+  firstLayoutCompleted() {
+    this.viewportCallback(this.isInViewport());
+
+    // Do not hide placeholder
+  }
+
+  /**
+   * @return {string}
+   **/
+  buildUrl_() {
+    const suffix = (this.random) ?
+        `/tokens/${this.mediaAttribute_}/interactions/random` :
+        `/interactions/${this.mediaAttribute_}/display`;
+    return encodeURIComponent(`${this.displayBaseUrl_}${suffix}`);
+  }
+
+  /**
+   * @return {!Promise<!JSONType>}
+   **/
+  queryMedia_() {
+    const url = decodeURIComponent(this.buildUrl_());
```

Comment:
Decode doesn't really make sense here, or maybe I'm not understanding this fully. See comments above.